### DISCUSSION TITLE
CustomComponent as indicator for mouseExit / mouseEnter

### DISCUSSION THREAD

**Discussion by Tonfisch** [RomanKubiak/ctrlr#566] [Q&A] [answered]
Hello,
I need help with this.
I want a fader to send a message "A" when a mouse enters it's boundarys, and message "B" when it leaves.
I thought a customComponent on top of the fader would be the trick, since it has the right callback funktions for mouseEnter / mouseExit. That works so far. The problem is that i don't understand how the customComponent works. If it is on top, i can't control the fader. If it's at buttom, the callbacks are not working. 
Is there someone who can help?

**Comment by damiensellier** [comment, marked as answer]
Hi there's some info about it here : 
https://github.com/RomanKubiak/ctrlr/issues/361#issuecomment-946477965

You need to create a custom component sitting below your slider with "component is at the back".
![Component is at the back](https://github.com/RomanKubiak/ctrlr/assets/55164385/0c67be63-1039-4a1d-9742-4fe577e389b1)

**Save your panel, close and open it again to redraw it with those position properties**.

You can use Juce triggers mouseEnters() and mouseExit()
https://docs.juce.com/master/classMouseListener.html#a479dc784712648193bdc236b2c34e8bd


Yeah, I agree, when you have two component stacked up, the custom component has to be bigger than the slider.
When you go over the custom component it triggers the mouseEnter() and this is what you'll need to reset the message or replace it with the "default" one. 

So the custom component  working only if you want to remove or replace the message with its callback "mouseEnter()"

Slider must be on top of the custom component.

uiSlider triggers the script to display your message with "Called when the mouse is moved over the component"
![MouseOver Callback](https://github.com/RomanKubiak/ctrlr/assets/55164385/99d1b05f-7b24-4dc6-80a4-2cbf5ded9f55)


uiCustomComponent removes or resets the message with ""MouseEnter callback".  It can work for different sliders since everytime you'll leave a slider area you'll activate the custom component with a "mouseEnter".

![Custom Component callbacks](https://github.com/RomanKubiak/ctrlr/assets/55164385/ef7d6fd1-755a-4433-a29c-68d1ac0bab13)


There's also other ways to remove or replace your message, for exemple with a timer.

If you don't need to reset the message when you leave the slider area, forget about the custom component's callbacks.

Hope it helps

Damien


EDIT:you'll find attached a demo for this little trick
[MessageResetOnCustomEnter_1_0_2023-07-30_02-30.panel.zip](https://github.com/RomanKubiak/ctrlr/files/12208638/MessageResetOnCustomEnter_1_0_2023-07-30_02-30.panel.zip)

**Reply by Tonfisch** [reply]
Hi Damian,
thanks for the demo and your ideas.
Yes, the trick is to use only one callback at the customComponent the second Info has to be the callback at the Slider.
The only drawback is that the callback  at the slider sends out many events (mouse over component). In my case it seems work.
Optimum would be only one event because i send Midi to the Device , for a push-turn encoder.
But that's whining on very high level(-:
Many thanks for your help

Btw. there is a funny thing going on when i set the callback-slots to none. It does dot stopp the callback from working. I have to put a dummy callback in place to stop it. Did that ever happen on you?

Toni

**Reply by damiensellier** [reply]
Nope, I don't have this bug I use CTRLR 5.6.1 for mac.
You can display the message with the mouse move-over callback and send the midi with the other event "Called when modulator value changes".
I know it sends the message like every 5ms when moving over the area, it would require adding the mouseEnter() and mouseExit() for the sliders at core level. I'll add it to my list for my custom ctrlr fork.
If you think your question has been replied, please mark the related post as the answer. 
Enjoy!

**Reply by dnaldoog** [reply]
> The only drawback is that the callback at the slider sends out many events (mouse over component). 

You can block the repeating many mouseover events using boolean truth:

Also in this panel I have put the component into a layer below the uiSlider. But either way is fine.

[MessageResetOnCustomEnter_2_0.zip](https://github.com/RomanKubiak/ctrlr/files/12297565/MessageResetOnCustomEnter_2_0.zip)


<pre>
mouseEnter = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    free = true -- must be global ie not local
    console("hide the message because mouse is out the slider zone")
    panel:getLabel("label"):setText("your mouse is outside the slider")
end

sliderOver = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    if free then
        console("display your message A because the mouse is over the slider")
        panel:getLabel("label"):setText("your mouse is on top of the slider")
    end
    free = false
end

</pre>

**Reply by damiensellier** [reply]
Hi John, It works great! I saw that you assigned the two components to 2 different layers. It's way better than the "component is at the back" property.

**Reply by dnaldoog** [reply]
Great to hear Damien!

**Reply by Tonfisch** [reply]
Hello John,
that works realy good!
I thought the top-layer would block all events for components that are on lower layers.
Are there other events coming through, or is that unic to the CustomComponent?

I like those illustrated panels very much.
Thank you very much
Toni

**Reply by dobo365** [reply]
This is indeed a great one that I added it to my "bible" ;-)
It is complementary to what I developed for drawing patch cables between patch points which is based on mouse click - drag and move - mouse release.
When the cables are going over sliders I also need to use a trick as the custom component is covering the sliders and make them not reachable. So, there is a copy of each slider on top of the bottom sliders. Only the bottom sliders have the image. The top sliders are handling the ValueChange and MouseClick methods. The value change method of the top slider is making the bottom slider move.

I'm still not using layers (John will laugh, we discussed layers and he made great explanations about them - also added in the bible).
I discovered that "Send to back" doesn't actually work and I'm not sure what it is doing.
**What is for sure taken into account (and this is what I'm controlling) is the order of the components in the panel xml file.**
So, you need to find first the bottom sliders then the custom component then the top sliders.
You can easily move each component related code using Notepad++ or similar without making any error.

**Reply by dnaldoog** [reply]
>I thought the top-layer would block all events for components that are on lower layers.
>Are there other events coming through, or is that unic to the CustomComponent?

Hi Toni,

Any object sitting on a layer beneath that is not obscured by an object in a layer above should be accessible (clickable etc) as far as I am aware. It's how I usually design panels personally and I have made panels with easily over 10 layers and found them to be stable (on 5.3.201) 😊

**Reply by Tonfisch** [reply]
Hi John,
the layers stop the show if you are in edit-mode. Not obvious for a beginner. (only for reference)
Do you know if other funktions are blocked in edit-mode?

**Reply by dnaldoog** [reply]
Hi Toni,

I think almost nothing works in edit mode, but I never test anything in edit mode anyway. Always I edit then test in panel mode. 🙂

**Comment by dnaldoog** [comment]
> I'm still not using layers (John will laugh, ... 🤣

That's interesting that it's the order in the XML that determines whether an object sits above another. I was going to say that I remember having trouble using the **send to back option**, but yes I always use layers anyway!
